File: 0338-counting-bits/0338-counting-bits.cpp
class Solution {
public:
    vector<int> countBits(int n) {

        vector<int> dp(n + 1, 0);
        for(int i = 1; i <= n; i++){
            // find the last bit if it is 1 or 0
            // reduce the curr no to half
            dp[i] = (i % 2) + dp[i / 2];
        }
        return dp;
    }
};

File: 0338-counting-bits/0338-counting-bits.java
class Solution {
    public int[] countBits(int n) {
        int[] ans = new int[n+1];
        ans[0] = 0;

        for(int i = 1; i <= n; i++){
            int setBits = 0;
            int m = i;
            while(m != 0){
                if((m & 1) == 1)
                    setBits++;
                m= m >> 1;
            }
            ans[i] = setBits;
        }
        return ans;
    }
}